Initial Functionality

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR implements the initial functionality for this Docker image. It provides a helper script that will either load a local CSV, or download the [GSA current Federal .gov domain list](https://github.com/GSA/data/blob/master/dotgov-domains/current-federal.csv), and then check each of the domains using the [cisagov/hash-http-content](https://github.com/cisagov/hash-http-content) package targeting the `/vulnerability-disclosure-policy` location. It will then output result CSVs for Agency-level and domain-level results.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

Although we are planning on deploying this functionality into AWS using lambdas, this will give a way for interested parties to run the core functionality on its own for testing/verification purposes.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I have run this against both the full CSV I mentioned above and a truncated list of the first 50 domains.
<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] Tests have been added and/or modified to cover the changes in this PR.
* [x] All new and existing tests pass.

mcdonnnj committed:
Update repository references

Replace skeleton-docker with vdp-scanner-docker and update the Lineage parent
repository.

mcdonnnj committed:
Update project references

Replace references to the example docker image with references to a vdp-scanner
image. Remove elements that will not be used in this Docker image.

mcdonnnj committed:
Create framework for image functionality

Create the framing of a vdp-scanner.py script that is the core functionality of
this Docker image. Adjust the Dockerfile and docker-compose.yml files to
reflect these changes.

mcdonnnj committed:
Add installation of serverless-chrome binary and dependencies

Update the Dockerfile and docker-compose.yml to reflect this addition. Rename
the scanner script to match preferred Python naming conventions.

mcdonnnj committed:
Implement vdp_scanner.py functionality

Implement the functionality to scan from a CSV of domains and output the Agency
and Domain level result CSVs. The README and compose file have been updated to
reflect these changes.

mcdonnnj committed:
Use pipenv for script requirements

This switches to using pipenv to manage the requirements for the vdp_scanner.py
script. This will help ensure deterministic Docker builds by installing
specific package versions, but stores them in a maintainable matter.

There was also a move back to using a root user because of issues with output
to the host machine and running under a different UID inside the Docker
container.

A debug option was added to the script and the version of chromium-common was
updated to resolve a build issue.

mcdonnnj committed:
Pare testing down to version checks

The nature of this Docker project does not lend itself well to runtime testing,
so for the time being I have pared testing down to version checks only.

dav3r commented on Dockerfile:
```diff
@@ -1,38 +1,68 @@
+# PIPENV_VENV_IN_PROJECT=1 directs pipenv to use the current directory for venvs
+RUN PIPENV_VENV_IN_PROJECT=1 pipenv sync
+
+RUN python -m pip uninstall --yes pipenv
```
Can you add a comment to briefly explain why you are uninstalling `pipenv`?

mcdonnnj replied:
Explanation added in ded105f.

dav3r commented on README.md:
```diff
@@ -1,14 +1,14 @@
+[![Docker Image Size (latest by date)](https://img.shields.io/docker/image-size/cisagov/vdp-scanner)](https://hub.docker.com/r/cisagov/vdp-scanner)
+[![Platforms](https://img.shields.io/badge/platforms-amd64%20%7C%20arm%2Fv6%20%7C%20arm%2Fv7%20%7C%20arm64%20%7C%20ppc64le%20%7C%20s390x-blue)](https://hub.docker.com/r/cisagov/vdp-scanner/tags)
 
 This is a docker skeleton project that can be used to quickly get a
```
Update this skeleton text to something more useful.

mcdonnnj replied:
I knew I was forgetting some of the little pieces. Updated in 040b430.

dav3r commented on docker-compose.yml:
```diff
@@ -3,35 +3,26 @@ version: "3.7"
-    secrets:
-      - source: quote_txt
-        target: quote.txt
+    user: "${UID}:${GID}"
```
Why are you doing this?  Probably worth a comment here.

mcdonnnj replied:
Ah, missed remnant from an approach I was trying. Removed in 5fb6b64.

jsf9k commented on README.md:
```diff
@@ -1,14 +1,14 @@
+[![Docker Image Size (latest by date)](https://img.shields.io/docker/image-size/cisagov/vdp-scanner)](https://hub.docker.com/r/cisagov/vdp-scanner)
+[![Platforms](https://img.shields.io/badge/platforms-amd64%20%7C%20arm%2Fv6%20%7C%20arm%2Fv7%20%7C%20arm64%20%7C%20ppc64le%20%7C%20s390x-blue)](https://hub.docker.com/r/cisagov/vdp-scanner/tags)
 
 This is a docker skeleton project that can be used to quickly get a
```
Can you update this blurb?

mcdonnnj replied:
Yep, totally forgot about getting back to that piece. Updated in 040b430.

jsf9k reviewed (approved):
I think this looks good.  I only ask for one piece of boilerplate in the `README.md` file to be updated.

dav3r commented on src/vdp_scanner.py:
```diff
@@ -0,0 +1,296 @@
+"""Check current federal DotGov domains for a Vulnerability Disclosure Policy.
```
Always good to define the acronym at the start.
```suggestion
"""Check current federal DotGov domains for a Vulnerability Disclosure Policy (VDP).
```

dav3r commented on src/vdp_scanner.py:
```diff
@@ -0,0 +1,296 @@
+            result = self._hasher.hash_url(urlunparse(url))
+        # If there is a TLS issue, try running it without verifying
+        except requests.exceptions.SSLError:
+            logging.warning("Trying '%s' without TLS verification", domain)
```
Use a consistent format for the fallback warnings
```suggestion
            logging.warning("Falling back to HTTPS without TLS verification for '%s'", domain)
```

dav3r commented on src/vdp_scanner.py:
```diff
@@ -0,0 +1,296 @@
+
+        self.agency_results[result.agency]["Total Domains"] += 1
+
+        if result.security_contact and result.security_contact != "(blank)":
```
Consider turning this magic value (`"(blank)"`) into a constant with an appropriate comment.

mcdonnnj replied:
Reasonable enough, updated in a739bed.

dav3r commented on src/vdp_scanner.py:
```diff
@@ -0,0 +1,296 @@
+            "--no-sandbox",
+            "--disable-gpu",
+            "--disable-dev-shm-usage",
+            "--no-zygote",
```
Consider adding a comment explaining why each browser option argument is being used here.

dav3r commented on src/vdp_scanner.py:
```diff
@@ -0,0 +1,296 @@
+    if args["--domain-csv"]:
+        scanner.domain_csv = args["--domain-csv"]
+
+    current_federal: List[Dict[str, str]]
```
I'd prefer to see a more informative variable name here than `current_federal`.

mcdonnnj replied:
I agree. It made sense when I first wrote the script only pull from the GSA list, but not so much now. Updated in 1388d68.

dav3r replied:
Much better, thanks! 👍

dav3r reviewed (approved):
Great work on this! 🦅 

Please consider my suggestions, but feel free to ignore any you strongly disagree with.

mcdonnnj committed:
Update README blurb describing this project

mcdonnnj committed:
Add comment explaining removal of pipenv

mcdonnnj committed:
Remove remnant from testing in docker-compose.yml

I did testing to use a supplied UID/GID to work around file ownership on output
CSVs, but I had issues with permissions inside the image and scrapped that
approach. I forgot a piece of that in the docker-compose.yml, so it needs to be
removed.

mcdonnnj committed:
Changes to scanner script from review

Define an acronym as part of the script's `__doc__` and change a logging output to be consistent with other logging.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Make the value for a missing security contact a class variable

This gives a useful name to refer to the value and allows it to be easily
changed without worrying about updating any references.

mcdonnnj committed:
Add comments explaining two more class variables

Add comments that explain what the agency_csv_header and domain_csv_header
variables are used for in the scanner script's VdpScanner class.

mcdonnnj committed:
Change to more descriptive variable name

The `current_federal` variable name is not particularly helpful with the
current functionality of the scanner script. I have updated it accordingly.

mcdonnnj committed:
Manually blacken vdp_scanner.py script

A merged suggestion on GitHub was not getting caught during local pre-commit
invocations. I manually ran black against the script to get the needed
reformatting to apply.

dav3r commented on README.md:
```diff
@@ -1,72 +1,56 @@
-appropriate for docker containers and the major languages that we use.
+This is a Docker project to scan either the
+[GSA current Federal .gov domain list](https://github.com/GSA/data/blob/master/dotgov-domains/current-federal.csv)
+, or a given CSV in the same format, with the
```
I'm nitpicking now, but the first comma is unnecessary here:
```suggestion
or a given CSV in the same format, with the
```

jsf9k replied:
The second comma is also unnecessary and awkward.
```suggestion
or a given CSV in the same format with the
```

mcdonnnj replied:
Alright you comma detractors, fixed in 025a9de.

dav3r replied:
Like, thanks, a, lot, man,

dav3r commented on README.md:
```diff
@@ -1,72 +1,56 @@
+[GSA current Federal .gov domain list](https://github.com/GSA/data/blob/master/dotgov-domains/current-federal.csv)
+, or a given CSV in the same format, with the
+[cisagov/hash-http-content](https://github.com/cisagov/hash-http-content)
+Python library. Then it will output CSVs with Agency and domain level results.
```
I don't think "agency" should be capitalized:
```suggestion
Python library. Then it will output CSVs with agency and domain level results.
```

mcdonnnj replied:
Fixed all unnecessary capitalization for that in 1ffbafa.

mcdonnnj committed:
Change case of agency and domain

Change to lowercase any uses of agency or domain that are not for CSV output
use.

mcdonnnj committed:
Remove unnecessary commas

It was pointed out that I was overusing commas in the README.

hillaryj commented:
It's working well when run on my local machine as well, except in the case where I lose internet or change network configuration settings (i.e. VPN) while it's running. I'll put the output from that into an issue (see https://github.com/cisagov/vdp-scanner-docker/issues/5), because it means to get the data out I have to completely restart the roughly 2-hr process (when run locally).

I'd like to see functionality tests as well as the container tests.

mcdonnnj committed:
Update openssl version installed and allow downgrades

Update to use the latest version of the openssl package. Add option to allow
apt to downgrade packages. This ensures that the specific versions we have
listed are what is installed.

mcdonnnj committed:
Remove linux/ppc64le from buildx platforms

The linux/ppc64le architecture does not have the chromium-common package. Since
this image relies on that package, we cannot support that architecture at this
time.

mcdonnnj committed:
Update Dockerfile package versions

Update the curl and openssl package versions to install the latest available.

mcdonnnj committed:
Remove linux/s390x from buildx platforms

The linux/s390x architecture does not have the chromium-common package. Since
this image relies on that package, we cannot support that architecture at this
time.

mcdonnnj committed:
Remove linux/arm/v6 from buildx platforms

The linux/arm/v6 architecture does not have the chromium-common package. Since
this image relies on that package, we cannot support that architecture at this
time.

mcdonnnj committed:
Update dependencies in Dockerfile

Add libxml2-dev and lxbslt1-dev packages and switch to using the non-slim
version of python Docker images in the compile stage to support successful
building of the lxml Python package on all platforms.

mcdonnnj committed:
Update Python dependencies for the Docker image

Updated by running `pipenv update` in the `src/` directory.

hillaryj reviewed (approved):
🌮
